polar-core/src/data_filtering.rs

```diff
@@ -1158,3 +1223,447 @@ mod test {
+                                        &self.types,
+                                        source,
+                                        lds.base.clone(),
+                                        vec![k.clone()],
+                                    )?;
+                                }
+                                _ => {}
+                            }
+                        }
+                        source =
+                            Self::joins(&self.types, source, lds.base.clone(), lds.path.clone())?;
+                        source =
+                            Self::joins(&self.types, source, rds.base.clone(), rds.path.clone())?;
+                        if let Source(lps) = &**lss {
+                            source = Self::joins(&self.types, source, lps.clone(), vec![])?;
+                        }
+                        if let Source(rps) = &**rss {
+                            source = Self::joins(&self.types, source, rps.clone(), vec![])?;
+                        }
+
+                        Ok(source)
+                    }
+
+                    (PreDatum::Field(Proj(lss, _), lds), _)
+                    | (_, PreDatum::Field(Proj(lss, _), lds)) => {
+                        let mut source = source?;
+                        source =
+                            Self::joins(&self.types, source, lds.base.clone(), lds.path.clone())?;
+                        if let Source(lps) = &**lss {
+                            source = Self::joins(&self.types, source, lps.clone(), vec![])?;
+                        }
+                        Ok(source)
+                    }
+                    _ => source,
+                }
+            },
+        )?;
+        let rc = self
+            .constraints
+            .into_iter()
+            .fold(Ok(source), |source, con| match con {
+                (PreDatum::Field(lhs, _), kind, rhs) | (rhs, kind, PreDatum::Field(lhs, _)) => {
+                    let source = source?;
+                    let rhs = rhs.into();
+                    Ok(Rc::new(Select {
+                        kind,
+                        lhs,
+                        rhs,
+                        source,
+                    }))
+                }
+                _ => err_invalid("QueryInfo::into_filter()".to_string()),
+            })?;
+        Ok((*rc).clone())
+    }
+
+    fn path_disasm(t: &Term) -> PolarResult<Vec<String>> {
+        use Operator::*;
+        match t.value() {
+            Value::Expression(Operation {
+                operator: Dot,
+                args,
+            }) if args.len() == 2 && args[1].value().as_string().is_ok() => {
+                let mut p = Self::path_disasm(&args[0])?;
+                p.push(args[1].value().as_string().unwrap().to_string());
+                Ok(p)
+            }
+            Value::Variable(Symbol(s)) => Ok(vec![s.clone()]),
+            Value::ExternalInstance(ExternalInstance {
```

Comment:
What's this handling?

Reply:
I think some case in GitClub that I had to hack around for demo purposes. Like the comment says, it's obviously wrong. External instances embedded in partials are usually bad news since they're actually ground values. Dot lookups on them should probably be dereferenced in the host before they get here.